types/nullable.go

```diff
@@ -0,0 +1,54 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+// nullBytes is a JSON null literal
+var nullBytes = []byte("null")
+
+// Nullable type which can help distinguish between if a value was explicitly
+// provided `null` in JSON or not
+type Nullable[T any] struct {
+	Value T
+	Set   bool
+	Null  bool
+}
+
+// UnmarshalJSON implements the Unmarshaler interface.
+func (t *Nullable[T]) UnmarshalJSON(data []byte) error {
+	t.Set = true
+	if bytes.Equal(data, nullBytes) {
+		t.Null = true
+		return nil
+	}
+	if err := json.Unmarshal(data, &t.Value); err != nil {
+		return fmt.Errorf("couldn't unmarshal JSON: %w", err)
+	}
+	t.Null = false
+	return nil
+}
+
+// MarshalJSON implements the  Marshaler interface.
+func (t Nullable[T]) MarshalJSON() ([]byte, error) {
+	if t.IsNull() {
+		return nullBytes, nil
```

Comment:
We need to replace this with:
```suggestion
		return []byte("null"), nil
```

Right now, we've got the risk that someone could do something like:

```go
func TestF(t *testing.T) {
	var n Nullable[string]
	n.Null = true
	b, err := n.MarshalJSON()
	fmt.Printf("b: %v\n", b)
	fmt.Printf("b: %s\n", b)
	fmt.Printf("err: %v\n", err)

	b[1] = '_'

	b_, err := n.MarshalJSON()
	fmt.Printf("b: %v\n", b_)
	fmt.Printf("b: %s\n", b_)
	fmt.Printf("err: %v\n", err)

}
```

Which outputs:

```
b: [110 117 108 108]
b: null
err: <nil>
b: [110 95 108 108]
b: n_ll
err: <nil>
```

Note that `nullBytes` from the package is changed.

We could alternatively perform a protective copy, or in this case, just return a literal

Reply:
Did not grok this in the first read. Trying to understand it.

Reply:
Ah! Makes sense now.. So basically some one can alter the nullByte. Good catch.